src/test/java/org/jenkinsci/plugins/workflow/steps/GeneralNonBlockingStepExecutionTest.java

```diff
@@ -115,6 +116,21 @@
         assertThat(logging.getRecords(), empty());
     }
 
+    @Issue("JENKINS-58878")
+    @Test public void shouldNotHang() throws Exception {
+        int iterations = 50;
+        startExit = new Semaphore(iterations); // Prevents the semaphores from blocking inside of the slowBlock step.
+        endExit = new Semaphore(iterations);
+        WorkflowJob p = r.createProject(WorkflowJob.class);
+        p.setDefinition(new CpsFlowDefinition(
+                "for (int i = 0; i < " + iterations + "; i++) {\n" +
+                "  slowBlock {\n" +
+                "    echo \"At ${i}\"\n" +
```

Comment:
I find

```suggestion
                "    echo(/At $i/)\n" +
```

a bit more legible.

Reply:
Improved in 707eaaf.